build/dsc/psmodule.rb

```diff
@@ -39,8 +49,11 @@ def attributes
     private
 
     def utf8_encode_content(content)
-      detection = CharlockHolmes::EncodingDetector.detect(content)
-      CharlockHolmes::Converter.convert content, detection[:encoding], 'UTF-8'
+      content.force_encoding(ENCODING)
+      unless content.valid_encoding?
+        content.encode!(ENCODING, ENCODING_OPTIONS)
```

Comment:
This overwrites the strings encoding with UTF-8, then tries to re-encode it as UTF-8 if not all byte sequences are valid, substituting replacement characters for byte sequences it doesn't understand... `valid_encoding?` is not a well named method because it doesn't really tell you if that string is actually really UTF-8 or not (which is a much more complex problem that charlock_holmes solves by leveraging ICU).

This won't work at all if the string is originally UCS-2 / UTF-16 or its something like ISO-8859-1, 1252, etc. Do we know if there's enforcement for PSD1 to be UTF-8 anywhere?

Note that the `File.open` above will default to using `Encoding.default_external` which is generally UTF-8 on Linux, but will be current codepage on Windows, which can lead to a world of hurt if local codepage is something like 932 and the file contains legit UTF-8 characters ;0

That said, we can solve this problem... I would call `File.read` with `utf-8|bom` on a first read attempt and maybe just look for `ModuleVersion` as you've alluded to, instead of reading all the attributes. If we can't find that we could try to fallback to reloading the file as `utf-16le|bom` and if we get no matches, abort reading the PSD1. We at least know all the characters involved in that string are ASCII and can make some concessions on how valid our file reading methodology is.

I can go into more detail if it's not clear what I'm suggesting.